## Homework
##### In this homework, we will use Credit Card Data from the previous homework.

* Note: sometimes your answer doesn't match one of the options exactly. That's fine. Select the option that's closest to your solution.

## Question 1
* Install Pipenv
* What's the version of pipenv you installed?
* Use --version to find out

In [3]:
!pipenv --version


pipenv, version 2022.10.9



## Question 2
* Use Pipenv to install Scikit-Learn version 1.0.2
* What's the first hash for scikit-learn you get in Pipfile.lock?
###### Note: you should create an empty folder for homework and do it there.

##### Models
##### We've prepared a dictionary vectorizer and a model.

##### They were trained (roughly) using this code:

In [ ]:
features = ['reports', 'share', 'expenditure', 'owner']
dicts = df[features].to_dict(orient='records')

dv = DictVectorizer(sparse=False)
X = dv.fit_transform(dicts)

model = LogisticRegression(solver='liblinear').fit(X, y)

#Note: You don't need to train the model. This code is just for your reference.

#And then saved with Pickle. Download them:

In [ ]:
!pip install wget

In [4]:
PREFIX='https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/course-zoomcamp/cohorts/2022/05-deployment/homework'

In [7]:
!python -m wget $PREFIX/model1.bin
!python -m wget $PREFIX/dv.bin


Saved under model1.bin

Saved under dv.bin


## Question 3
##### Let's use these models!

* Write a script for loading these models with pickle
* Score this client:

In [ ]:
#{"reports": 0, "share": 0.001694, "expenditure": 0.12, "owner": "yes"}

In [4]:
import pickle

model_file = 'model1.bin'
dv_file = 'dv.bin'

with open(model_file,'rb') as f:
    model = pickle.load(f)
    
with open(dv_file,'rb') as f:
    dv = pickle.load(f)

client = {"reports": 0, "share": 0.001694, "expenditure": 0.12, "owner": "yes"}

X = dv.transform([client])
y = model.predict_proba(X)

print(y)



[[0.83786586 0.16213414]]


##### Answer :  0.162

## Question 4
##### Now let's serve this model as a web service

* Install Flask and gunicorn (or waitress, if you're on Windows)
* Write Flask code for serving the model
* Now score this client using requests:

In [ ]:
# url = "YOUR_URL"
# client = {"reports": 0, "share": 0.245, "expenditure": 3.438, "owner": "yes"}
# requests.post(url, json=client).json()

In [5]:
client = {"reports": 0, "share": 0.245, "expenditure": 3.438, "owner": "yes"}

X = dv.transform([client])
y = model.predict_proba(X)

print(y)

[[0.0717782 0.9282218]]


##### Answer :  0.162

## Docker
##### Install Docker. We will use it for the next two questions.

##### For these questions, we prepared a base image: svizor/zoomcamp-model:3.9.12-slim. You'll need to use it (see Question 5 for an example).

##### This image is based on python:3.9.12-slim and has a logistic regression model (a different one) as well a dictionary vectorizer inside.

##### This is how the Dockerfile for this image looks like:

In [ ]:
FROM python:3.9.12-slim
WORKDIR /app
COPY ["model2.bin", "dv.bin", "./"]

In [2]:
!docker images

REPOSITORY              TAG           IMAGE ID       CREATED          SIZE
docker101tutorial       latest        4f696c9a586a   18 minutes ago   28.9MB
alpine/git              latest        b80d2cac43e4   2 days ago       43.6MB
svizor/zoomcamp-model   3.9.12-slim   571a6fdc554b   7 days ago       125MB


In [ ]:
#https://github.com/alexeygrigorev/mlbookcamp-code/blob/master/course-zoomcamp/cohorts/2022/05-deployment/homework.md

## Dockerfile
##### Now create your own Dockerfile based on the image we prepared.

##### It should start like that:

In [ ]:
FROM svizor/zoomcamp-model:3.9.12-slim
# add your stuff here

## Now complete it:

* Install all the dependencies form the Pipenv file
* Copy your Flask script
* Run it with Gunicorn
##### After that, you can build your docker image.

## Question 6
##### Let's run your docker container!

##### After running it, score this client once again:

In [ ]:
url = "YOUR_URL"
client = {"reports": 0, "share": 0.245, "expenditure": 3.438, "owner": "yes"}
requests.post(url, json=client).json()

##### What's the probability that this client will get a credit card now?


In [9]:
client = {"reports": 0, "share": 0.245, "expenditure": 3.438, "owner": "yes"}

X = dv.transform([client])
y = model.predict_proba(X)

print(y)

[[0.0717782 0.9282218]]


In [ ]:
docker pull svizor/zoomcamp-model:3.9.12-slim -> cmd
!docker images -> Jupyter
